pkg/platform/openstack/openstack-heat.go

```diff
@@ -404,37 +405,58 @@ func IsUserDataSame(ctx context.Context, userdata1, userdata2 string) bool {
 }
 
 // populateParams fills in some details which cannot be done outside of heat
-func (o *OpenstackPlatform) populateParams(ctx context.Context, VMGroupOrchestrationParams *vmlayer.VMGroupOrchestrationParams, action string) (*ReservedResources, error) {
+func (o *OpenstackPlatform) populateParams(ctx context.Context, VMGroupOrchestrationParams *vmlayer.VMGroupOrchestrationParams, action string) error {
 	log.SpanLog(ctx, log.DebugLevelInfra, "populateParams", "VMGroupOrchestrationParams", VMGroupOrchestrationParams.GroupName, "action", action)
 
-	// lock the resource reservations
-	resourceLock.Lock()
-	defer resourceLock.Unlock()
-	var reserved ReservedResources
-
-	usedCidrs := make(map[string]string)
 	if VMGroupOrchestrationParams.Netspec == nil {
-		return nil, fmt.Errorf("Netspec is nil")
+		return fmt.Errorf("Netspec is nil")
 	}
 	masterIP := ""
 	masterIPv6 := ""
 
-	if len(VMGroupOrchestrationParams.Subnets) > 0 {
+	ownerID := VMGroupOrchestrationParams.OwnerID
+	// reserve function may run several times if there are conflicts,
+	// so track which CIDRs need to be updated so they can be re-assigned
+	// on subsequent runs.
+	subnetNeedsResource := make(map[int]struct{})
+	for i, s := range VMGroupOrchestrationParams.Subnets {
+		if s.CIDR == vmlayer.NextAvailableResource {
+			subnetNeedsResource[i] = struct{}{}
```

Comment:
Do we need just one here? We can probably break after finding the first available.

Reply:
This is just tracking which subnets need to be assigned a reservable resource. The issue here is that the resource allocation function overwrites the "NextAvailableResource" value when it assigns a resource. But then that function may fail to commit, in which case that function gets run again. But the second time around, s.CIDR's value may not be "NextAvailableResource", it may be what was assigned on the previous run that did not commit. So here we're just tracking which subnets need CIDR's assigned so that the reservation function can run multiple times ok.

Reply:
Ah ok, now that makes sense.